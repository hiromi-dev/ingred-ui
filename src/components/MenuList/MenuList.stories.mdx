import { Meta, Story, ArgsTable, Canvas } from "@storybook/addon-docs/blocks";
import MenuList from "./MenuList";

<Meta title="Components/Navigation/MenuList" component={MenuList} />

# MenuList

MenuList is a lower-level component that is leveraged [`<Menu />`](/?path=/docs/components-navigation-menu--example).

<ArgsTable of={MenuList} />

## Samples

<Canvas>
  <Story
    name="Example"
    args={{
      contents: [
        { text: "Save", onClick: () => {}, type: "default" },
        { text: "Edit", onClick: () => {}, type: "default" },
        { text: "Delete", onClick: () => {}, type: "disabled" },
        { text: "Update", onClick: () => {}, type: "warning" },
      ],
      maxHeight: "100px",
    }}
  >
    {(args) => (
      <div style={{ backgroundColor: "silver", padding: "10px" }}>
        <MenuList {...args} />
      </div>
    )}
  </Story>
</Canvas>
