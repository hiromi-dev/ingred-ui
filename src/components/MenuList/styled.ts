import styled from "styled-components";
import { Property } from "csstype";
import { addScrollbarProperties } from "../../utils/scrollbar";
import Typography from "../Typography";

type ContainerProps = {
  inline: boolean;
  maxHeight: Property.MaxHeight;
};

export const Container = styled.div<ContainerProps>`
  display: ${({ inline }) => (inline ? "inline-block" : "block")};
  padding: ${({ theme }) => theme.spacing}px 0;
  border-radius: ${({ theme }) => theme.radius}px;
  background-color: ${({ theme }) => theme.palette.background.default};
  ${({ maxHeight }) => addScrollbarProperties({ maxHeight })}
`;

export const Text = styled(Typography)``;

export const TextContainer = styled.div<{
  disabled: boolean;
  bgColorAtHover: string;
  bgColorAtActive: string;
  textColorAtHover: string;
}>`
  cursor: ${({ disabled }) => (disabled ? "default" : "pointer")};
  display: flex;
  align-items: center;
  height: 32px;
  margin: 0 ${({ theme }) => theme.spacing}px;
  padding: 0 ${({ theme }) => theme.spacing}px;
  border-radius: ${({ theme }) => theme.radius}px;
  &:hover {
    ${Text} {
      color: ${({ textColorAtHover }) => textColorAtHover};
    }
    background-color: ${({ bgColorAtHover }) => bgColorAtHover};
  }
  &:active {
    background-color: ${({ bgColorAtActive }) => bgColorAtActive};
  }
`;
