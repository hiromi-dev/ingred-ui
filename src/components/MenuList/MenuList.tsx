import React from "react";
import * as Styled from "./styled";
import { Property } from "csstype";
import Divider from "../Divider";
import Typography from "../Typography";
import { useTheme } from "../../themes";

export type ContentProp = React.ComponentPropsWithRef<"div"> & {
  text: string;
  onClick: () => void;
  divideTop?: boolean;
  disabled?: boolean;
};

export type GroupContentProp = React.ComponentPropsWithRef<"div"> & {
  title: string;
  contents: ContentProp[];
};

export type MenuListProps = React.ComponentPropsWithRef<"div"> & {
  inline?: boolean;
  contents: ContentProp[] | GroupContentProp[];
  maxHeight?: Property.MaxHeight;
};

const MenuList = React.forwardRef<HTMLDivElement, MenuListProps>(
  ({ inline = false, contents, maxHeight = "none", ...rest }, ref) => {
    const theme = useTheme();

    const renderGroupContent = (content: GroupContentProp) => (
      <>
        <Styled.TitleContainer>
          <Typography size="sm">{content.title}</Typography>
        </Styled.TitleContainer>
        {content.contents.map((content: ContentProp) => (
          <Styled.TextContainer
            key={content.title}
            disabled={content.disabled}
            onClick={content.onClick}
          >
            <Typography
              size="sm"
              color={content.disabled ? "disabled" : "initial"}
            >
              {content.text}
            </Typography>
          </Styled.TextContainer>
        ))}
      </>
    );

    const renderContent = (content: ContentProp) => (
      <React.Fragment key={content.text}>
        {content.divideTop && (
          <Divider my={1} mx={2} color={theme.palette.gray.light} />
        )}
        <Styled.TextContainer
          disabled={content.disabled}
          onClick={content.onClick}
        >
          <Typography
            size="sm"
            color={content.disabled ? "disabled" : "initial"}
          >
            {content.text}
          </Typography>
        </Styled.TextContainer>
      </React.Fragment>
    );

    return (
      <Styled.Container
        inline={inline}
        maxHeight={maxHeight}
        {...rest}
        ref={ref}
      >
        {contents.map((content: ContentProp | GroupContentProp) =>
          content.title
            ? renderGroupContent(content as GroupContentProp)
            : renderContent(content as ContentProp),
        )}
      </Styled.Container>
    );
  },
);

export default MenuList;
