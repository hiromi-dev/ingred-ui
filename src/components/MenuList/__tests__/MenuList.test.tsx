import * as React from "react";
import "@testing-library/jest-dom/extend-expect";
import { cleanup } from "@testing-library/react";
import MenuList from "..";
import { renderWithThemeProvider } from "../../../utils/renderWithThemeProvider";
import { ContentProp } from "../MenuList";

const contents: ContentProp[] = [
  {
    text: "Save",
    onClick: () => {},
    type: "default",
  },
  {
    text: "Save and execute",
    onClick: () => {},
    divideTop: true,
    type: "default",
  },
  {
    text: "Save as draft",
    onClick: () => {},
    type: "warning",
  },
  {
    text: "Cancel",
    onClick: () => {},
    type: "disabled",
  },
];

describe("MenuList component testing", () => {
  afterEach(cleanup);

  test("MenuList splited", () => {
    const { asFragment } = renderWithThemeProvider(
      <MenuList contents={contents} />,
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
