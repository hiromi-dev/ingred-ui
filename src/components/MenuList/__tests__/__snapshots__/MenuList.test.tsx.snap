// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MenuList component testing MenuList splited 1`] = `
<DocumentFragment>
  <div
    class="sc-bdfBwQ hWeOum"
  >
    <hr
      class="sc-dlfnbm jOeRSr"
      color="#E2E8EA"
      orientation="horizontal"
    />
    <div
      class="sc-gsTCUz jkzjCh"
    >
      <p
        class="sc-hKgILt jKKuBF"
        color="#041C33"
        font-size="13px"
      >
        保存する
      </p>
    </div>
    <div
      class="sc-gsTCUz jkzjCh"
    >
      <p
        class="sc-hKgILt jKKuBF"
        color="#041C33"
        font-size="13px"
      >
        保存して実行する
      </p>
    </div>
    <hr
      class="sc-dlfnbm jOeRSr"
      color="#E2E8EA"
      orientation="horizontal"
    />
    <div
      class="sc-gsTCUz jkzjCh"
    >
      <p
        class="sc-hKgILt jKKuBF"
        color="#041C33"
        font-size="13px"
      >
        下書きとして保存する
      </p>
    </div>
    <div
      class="sc-gsTCUz jkzjCh"
    >
      <p
        class="sc-hKgILt jKKuBF"
        color="#041C33"
        font-size="13px"
      >
        やっぱり何もしない
      </p>
    </div>
  </div>
</DocumentFragment>
`;
